File: cli/internal/client/channels.go
package client

import (
	"context"
	"net/url"
)

// Channel represents one row from /api/channels.
type Channel struct {
	ID            string                 `json:"id"`
	Profile       string                 `json:"profile"`
	ChannelType   string                 `json:"channel_type"`
	Mode          string                 `json:"mode"`
	AuthMode      string                 `json:"auth_mode"`
	ResponseMode  string                 `json:"response_mode"`
	Enabled       bool                   `json:"enabled"`
	Status        string                 `json:"status"`
	Config        map[string]any         `json:"config"`
	State         map[string]any         `json:"state"`
	CreatedAt     float64                `json:"created_at"`
	UpdatedAt     float64                `json:"updated_at"`
}

// CreateChannelRequest is the body for POST /api/channels.
type CreateChannelRequest struct {
	ChannelType  string         `json:"channel_type"`
	Mode         string         `json:"mode,omitempty"`
	AuthMode     string         `json:"auth_mode,omitempty"`
	ResponseMode string         `json:"response_mode,omitempty"`
	Enabled      *bool          `json:"enabled,omitempty"`
	Config       map[string]any `json:"config,omitempty"`
}

// UpdateChannelRequest is the body for PATCH /api/channels/{id}. All fields optional.
type UpdateChannelRequest struct {
	Mode         *string        `json:"mode,omitempty"`
	AuthMode     *string        `json:"auth_mode,omitempty"`
	ResponseMode *string        `json:"response_mode,omitempty"`
	Enabled      *bool          `json:"enabled,omitempty"`
	Config       map[string]any `json:"config,omitempty"`
}

// ListChannels returns all channels for the active profile.
func (c *Client) ListChannels(ctx context.Context) ([]Channel, error) {
	var resp struct {
		Channels []Channel `json:"channels"`
	}
	if err := c.GetJSON(ctx, "/api/channels", &resp); err != nil {
		return nil, err
	}
	return resp.Channels, nil
}

// GetChannelCatalog returns the TOML-defined catalog (per-type metadata).
func (c *Client) GetChannelCatalog(ctx context.Context) (map[string]any, error) {
	var resp struct {
		Channels map[string]any `json:"channels"`
	}
	if err := c.GetJSON(ctx, "/api/channels/catalog", &resp); err != nil {
		return nil, err
	}
	return resp.Channels, nil
}

// CreateChannel registers a new channel; returns the created row.
func (c *Client) CreateChannel(ctx context.Context, req CreateChannelRequest) (*Channel, error) {
	var resp struct {
		Channel Channel `json:"channel"`
	}
	if err := c.PostJSON(ctx, "/api/channels", req, &resp); err != nil {
		return nil, err
	}
	return &resp.Channel, nil
}

// UpdateChannel patches a channel.
func (c *Client) UpdateChannel(ctx context.Context, id string, req UpdateChannelRequest) (*Channel, error) {
	var resp struct {
		Channel Channel `json:"channel"`
	}
	if err := c.PatchJSON(ctx, "/api/channels/"+url.PathEscape(id), req, &resp); err != nil {
		return nil, err
	}
	return &resp.Channel, nil
}

// DeleteChannel removes a channel (cascades conversations + senders).
func (c *Client) DeleteChannel(ctx context.Context, id string) error {
	return c.Delete(ctx, "/api/channels/"+url.PathEscape(id), nil)
}

// ChannelAuthEventsPath returns the path for the channel's pairing-event SSE
// stream — used by `opa channels pair` to render the QR / prompt for codes.
func (c *Client) ChannelAuthEventsPath(id string) string {
	return "/api/channels/" + url.PathEscape(id) + "/auth-events"
}

// SubmitChannelAuthInput POSTs a verification code or 2FA password back to a
// channel that's in the middle of an interactive pairing flow.
//
// Either ``code`` or ``password`` should be non-empty. The server returns
// ``409 No auth input expected`` when the adapter isn't waiting for input
// (most often because pairing already completed).
func (c *Client) SubmitChannelAuthInput(ctx context.Context, id, code, password string) error {
	body := map[string]string{}
	if code != "" {
		body["code"] = code
	}
	if password != "" {
		body["password"] = password
	}
	return c.PostJSON(ctx, "/api/channels/"+url.PathEscape(id)+"/auth-input", body, nil)
}

File: app/cli/client/channels.py
"""Channels endpoints — `/api/channels*`.

Mirrors `cli/internal/client/channels.go`. The pairing SSE stream is consumed
by the `channels pair` command (Phase 4); list/catalog/create/delete are
straightforward JSON.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Optional
from urllib.parse import quote

from app.cli.client._base import Client


@dataclass(frozen=True)
class Channel:
    id: str
    profile: str
    channel_type: str
    mode: str
    auth_mode: str
    response_mode: str
    enabled: bool
    status: str
    config: dict[str, Any] = field(default_factory=dict)
    state: dict[str, Any] = field(default_factory=dict)
    created_at: float = 0.0
    updated_at: float = 0.0

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "Channel":
        config = d.get("config") if isinstance(d.get("config"), dict) else {}
        state = d.get("state") if isinstance(d.get("state"), dict) else {}
        return cls(
            id=str(d.get("id") or ""),
            profile=str(d.get("profile") or ""),
            channel_type=str(d.get("channel_type") or ""),
            mode=str(d.get("mode") or ""),
            auth_mode=str(d.get("auth_mode") or ""),
            response_mode=str(d.get("response_mode") or ""),
            enabled=bool(d.get("enabled") or False),
            status=str(d.get("status") or ""),
            config=config or {},
            state=state or {},
            created_at=float(d.get("created_at") or 0),
            updated_at=float(d.get("updated_at") or 0),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "profile": self.profile,
            "channel_type": self.channel_type,
            "mode": self.mode,
            "auth_mode": self.auth_mode,
            "response_mode": self.response_mode,
            "enabled": self.enabled,
            "status": self.status,
            "config": self.config,
            "state": self.state,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }


async def list_channels(client: Client) -> list[Channel]:
    resp = await client.get_json("/api/channels")
    if isinstance(resp, dict) and isinstance(resp.get("channels"), list):
        return [Channel.from_dict(c) for c in resp["channels"] if isinstance(c, dict)]
    return []


async def get_channel_catalog(client: Client) -> dict[str, Any]:
    resp = await client.get_json("/api/channels/catalog")
    if isinstance(resp, dict) and isinstance(resp.get("channels"), dict):
        return resp["channels"]
    return {}


async def create_channel(
    client: Client,
    *,
    channel_type: str,
    mode: str = "",
    auth_mode: str = "",
    response_mode: str = "",
    enabled: Optional[bool] = None,
    config: Optional[dict[str, Any]] = None,
) -> Channel:
    body: dict[str, Any] = {"channel_type": channel_type}
    if mode:
        body["mode"] = mode
    if auth_mode:
        body["auth_mode"] = auth_mode
    if response_mode:
        body["response_mode"] = response_mode
    if enabled is not None:
        body["enabled"] = enabled
    if config:
        body["config"] = config
    resp = await client.post_json("/api/channels", body)
    if isinstance(resp, dict) and isinstance(resp.get("channel"), dict):
        return Channel.from_dict(resp["channel"])
    raise RuntimeError("unexpected /api/channels response")


async def delete_channel(client: Client, channel_id: str) -> None:
    await client.delete(f"/api/channels/{quote(channel_id, safe='')}")


def channel_auth_events_path(channel_id: str) -> str:
    return f"/api/channels/{quote(channel_id, safe='')}/auth-events"


async def submit_channel_auth_input(
    client: Client,
    channel_id: str,
    code: str = "",
    password: str = "",
) -> None:
    body: dict[str, str] = {}
    if code:
        body["code"] = code
    if password:
        body["password"] = password
    await client.post_json(
        f"/api/channels/{quote(channel_id, safe='')}/auth-input",
        body,
    )
